Improve numerical stability of PhaseQuadMuon

jclarkeSTFC opened the pull request:
This algorithm was inverting a 2x2 matrix manually, which resulted in numerically unstable results. On different CPU architectures, the results are wildly different, and even doing the calculation in a slightly different order changes the results. By using `EigenMatrix`, we get a more stable inverse.

#### Purpose of work

`PhaseQuadMuon` needs to give consistent results on different CPU architectures. This is important now that we are making an ARM build of Mantid, see #38990.

*There is no associated issue.*

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jclarkeSTFC committed:
Improve numerical stability of PhaseQuadMuon

This algorithm was inverting a 2x2 matrix manually, which resulted in
numerically unstable results. On different CPU architectures, the
results are wildly different, and even doing the calculation in a
slightly different order changes the results. By using EigenMatrix, we
get a more stable inverse.

jclarkeSTFC committed:
Added a release note

adriazalvarez commented on Framework/Muon/src/PhaseQuadMuon.cpp:
```diff
@@ -237,44 +238,47 @@ API::MatrixWorkspace_sptr PhaseQuadMuon::squash(const API::MatrixWorkspace_sptr
+  muLamMatrix.set(1, 1, syy);
+  muLamMatrix.invert();
+
+  std::vector<double> aj(nspec), bj(nspec);
+  for (size_t h = 0; h < nspec; h++) {
+    if (emptySpectrum[h]) {
+      aj[h] = 0;
+      bj[h] = 0;
+    } else {
+      const auto factors = muLamMatrix * n0Vectors[h];
+      aj[h] = factors[0];
+      bj[h] = factors[1];
```
```suggestion
  std::vector<double> aj(nspec), bj(nspec);
  for (size_t h = 0; h < nspec; h++) {
  aj[h] = bj[h] = 0;
    if (!emptySpectrum[h]) {
      const auto factors = muLamMatrix * n0Vectors[h];
      aj[h] = factors[0];
      bj[h] = factors[1];

```
I like a little bit more this style (same with previous if/else), but feel free to leave as it is if you disagree or find this heretic.

adriazalvarez commented on Framework/Muon/test/PhaseQuadMuonTest.h:
```diff
@@ -223,19 +223,19 @@ class PhaseQuadMuonTest : public CxxTest::TestSuite {
     const auto specImE = outputWs->getSpectrum(1).e();
     // Check real Y values
-    TS_ASSERT_DELTA(specReY[0], 2.1969, 0.0001);
+    TS_ASSERT_DELTA(specReY[0], 2.3212, 0.0001);
```
Delta for all tests is 0.0001. 
- Could we add a constexpr DELTA to the tests instead of putting it in every assertion? 
- Is this delta number arbitrary? Since you have change the way to perform operations to converge numerical precision in different system, are we expecting delta between systems to be less than this, or we can decrease the number?

jclarkeSTFC replied:
I've added `constexpr` for the tolerance. To check if the tolerance can be reduced would require some testing on the different operating systems and CPU architectures, so I haven't done that.

adriazalvarez commented on Framework/Muon/test/PhaseQuadMuonTest.h:
```diff
@@ -223,19 +223,19 @@ class PhaseQuadMuonTest : public CxxTest::TestSuite {
-    TS_ASSERT_DELTA(specImY[0], -0.1035, 0.0001);
-    TS_ASSERT_DELTA(specImY[20], -0.0006, 0.0001);
-    TS_ASSERT_DELTA(specImY[50], 0.0047, 0.0001);
+    TS_ASSERT_DELTA(specImY[0], 0.0328, 0.0001);
+    TS_ASSERT_DELTA(specImY[20], -0.0003, 0.0001);
+    TS_ASSERT_DELTA(specImY[50], -0.0033, 0.0001);
     // Check imaginary E values
```
It seems the specImY[n] parameters is the one with the most variations in this new way of processing the inverse, is there a reason for this ?

adriazalvarez reviewed (changes requested):
Changes look fine to me. Adding the eigenvector class make it easier to see the intentions of the vector manipulations. 
I have some minor question/suggestions.

jclarkeSTFC committed:
Improve logic clarity

Co-authored-by: Adri Diaz <146007827+adriazalvarez@users.noreply.github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jclarkeSTFC committed:
Use constant for comparison delta

adriazalvarez reviewed (changes requested):
The inconsistencies between numerical data in the tests are a bit strange.
Running the example code from https://docs.mantidproject.org/nightly/algorithms/PhaseQuad-v1.html I get all nans on the output workspace. I'm not sure, but it looks like eigen libraries can't handle very big numbers. 
As an alternative, can we try calculating the inverse using the [Matrix class in Mantid Kernel instead?](https://github.com/mantidproject/mantid/blob/main/Framework/Kernel/inc/MantidKernel/Matrix.h)  This way we can also avoid linking curve fitting to muon.

adriazalvarez commented:
> > The inconsistencies between numerical data in the tests are a bit strange. Running the example code from https://docs.mantidproject.org/nightly/algorithms/PhaseQuad-v1.html I get all nans on the output workspace. I'm not sure, but it looks like eigen libraries can't handle very big numbers. As an alternative, can we try calculating the inverse using the [Matrix class in Mantid Kernel instead?](https://github.com/mantidproject/mantid/blob/main/Framework/Kernel/inc/MantidKernel/Matrix.h) This way we can also avoid linking curve fitting to muon.
> 
> I tried using `Kernel::Matrix` and it works, but the results on ARM are slightly different, e.g. 1.65 vs 3.073 for one of the comparisons. So I could do that and change the tolerances.
> 
> When I ran the example code from the docs with the `Matrix` class, the determinant is close to zero, so I suspect that's where the NaNs came from with `EigenMatrix`, not a great choice of example to have a situation without a well defined solution.


I see. If you feel ok with using eigen matrix I don't want to keep beating around the bushes, perhaps we could just use a different example file in the docs with a `better` solution. There are few muon files in the example dataset. Some of them could be a bit better?

jclarkeSTFC committed:
Change docs example to have well-defined solution

This arbitrary example in the docs was not well-defined. The solution
required inverting a matrix that was close to being singular, resulting
in unreliable answers.

jclarkeSTFC commented:
> > > The inconsistencies between numerical data in the tests are a bit strange. Running the example code from https://docs.mantidproject.org/nightly/algorithms/PhaseQuad-v1.html I get all nans on the output workspace. I'm not sure, but it looks like eigen libraries can't handle very big numbers. As an alternative, can we try calculating the inverse using the [Matrix class in Mantid Kernel instead?](https://github.com/mantidproject/mantid/blob/main/Framework/Kernel/inc/MantidKernel/Matrix.h) This way we can also avoid linking curve fitting to muon.
> > 
> > 
> > I tried using `Kernel::Matrix` and it works, but the results on ARM are slightly different, e.g. 1.65 vs 3.073 for one of the comparisons. So I could do that and change the tolerances.
> > When I ran the example code from the docs with the `Matrix` class, the determinant is close to zero, so I suspect that's where the NaNs came from with `EigenMatrix`, not a great choice of example to have a situation without a well defined solution.
> 
> I see. If you feel ok with using eigen matrix I don't want to keep beating around the bushes, perhaps we could just use a different example file in the docs with a "better" solution. There are few muon files in the example dataset. Some of them could be a bit better?

I've edited the docs example to not be undefined, the detector info is arbitrary, so I altered the second set of phases slightly.
